### PR TITLE
Refactor dropzone for unmodified default blocks with `useBlockDropZone` and `InsertionPoint`

### PR DESCRIPTION
**Pull request by kevin940726**
## What?

A follow-up and refactor of #42722, an alternative to #44606.

This refactor basically reverts the implementation in the Paragraph block (#42722) and re-implements the same feature in `useBlockDropZone` and `<InsertionPoint>` using 
`<BlockPopover>` and `isUnmodifiedDefaultBlock` (as suggested in https://github.com/WordPress/gutenberg/pull/42722#issuecomment-1262025220). This refactor allows us to reuse the same drop zone and their event listeners so that we don't have to bind them elsewhere.

## Why?

The original PR fails some performance metrics when rendering a large list of empty paragraph blocks and it doesn't take `setDefaultBlock` into account. This PR solves both issues.

## How?

1. We calculate the drop indicator and take unmodified default blocks into account in `useBlockDropZone`.
2. If there's an unmodified default block nearby then change the drop `operation` to `replace` and call `showInsertionPoint` with it.
3. Render the dropzone indicator (fully presentational component) on top of the unmodified default block using `<BlockPopover>`.

This PR has a nice side effect that dragging in-between unmodified default blocks will also trigger the drop zone indicator and allow dropping. This is smoother than the previous implementation, where only hovering on the empty paragraph blocks will trigger the drop zone (which is often very small). See it in action in the below screencast.

## Testing Instructions

1. Create some empty paragraph blocks (unmodified default blocks) and some other blocks with contents.
2. Drag something around the post and see that the dropzone indicator appears when hovering on the empty paragraph blocks.
3. Drop to replace the empty paragraph blocks with some blocks (depending on the dragged data).

## Screenshots or screencast 

https://user-images.githubusercontent.com/7753001/193731962-73874eca-a7ef-42bb-b06c-714388b507c8.mp4

## How to review

1. See `useBlockDropZone` first, and follow the logic there.
2. `block-editor`'s actions (`showInsertionPoint`) -> reducer.
3. `<InsertionPoint>` is where the presentational component renders.
4. `<BlockPopoverDropZone>` (`block-popover/drop-zone.js`) is where the presentational drop zone component is.
5. Other source files are mostly code removals from #42722.
6. The rests are tests and renamings.

### CODE REVIEW

File: packages/block-editor/src/components/use-block-drop-zone/index.js [hunk 1]
```diff
@@ -174,12 +237,10 @@ export default function useBlockDropZone( {
 		onDragLeave() {
 			throttled.cancel();
 			hideInsertionPoint();
-			setTargetBlockIndex( null );
```
Comment (kevin940726) [inline packages/block-editor/src/components/use-block-drop-zone/index.js:177]: I'm actually not sure why we needed to set it to `null` here as it'll get updated on drag over anyway?

This is making the hook returned by `useOnBlockDrop` receive the wrong index sometimes.

File: packages/block-editor/src/components/use-block-drop-zone/test/index.js [hunk 2]
```diff
@@ -130,46 +141,10 @@ describe( 'getNearestBlockIndex', () => {
 				orientation
 			);
 
-			expect( result ).toBe( 2 );
-		} );
-
-		it( 'returns `2` when the position is nearest to the start of the third block', () => {
```
Comment (kevin940726) [inline packages/block-editor/src/components/use-block-drop-zone/test/index.js:136]: I deleted some tests here as I think they are a bit repetitive. @talldan is there something that I missed but is important in those tests?

File: packages/block-library/CHANGELOG.md [hunk 3]
```diff
@@ -9,7 +9,6 @@
 ### New Feature
 
 -   Made it possible to import individual blocks ([#42258](https://github.com/WordPress/gutenberg/pull/42258)). Check [README](./README.md#loading-individual-blocks) for more information.
--   Paragraph block: You can now drop files/blocks/HTML on an empty Paragraph block to transform it into relevant blocks ([#42722](https://github.com/WordPress/gutenberg/pull/42722)).
```
Comment (kevin940726) [inline packages/block-library/CHANGELOG.md:14]: It accidentally went into the wrong release changelog while it hasn't been released yet. I think it's okay to remove it?

File: packages/block-editor/src/components/use-block-drop-zone/index.js (hunk 1 above)
Comment (kevin940726) [reply packages/block-editor/src/components/use-block-drop-zone/index.js:177]: Note that there is some flickering in Safari when dragging in between two blocks:

https://user-images.githubusercontent.com/7753001/193735124-9d689e41-f8fa-44c8-932f-59aab71f8112.mp4

It's because Safari will always [return `null` for `event.relatedTarget` in the `dragleave` events](https://bugs.webkit.org/show_bug.cgi?id=66547). I think there are ways to fix it but probably deserves another PR.

File: packages/block-editor/src/components/use-block-drop-zone/index.js [hunk 4]
```diff
@@ -43,48 +54,79 @@ export function getNearestBlockIndex( elements, position, orientation ) {
...
+
+		let [ distance, edge ] = getDistanceToNearestEdge(
 			position,
 			rect,
 			allowedEdges
 		);
+		// Prioritize the element if the point is inside of it.
+		if ( isPointContainedByRect( position, rect ) ) {
+			distance = 0;
+		}
```
Comment (talldan) [inline packages/block-editor/src/components/use-block-drop-zone/index.js]: This is an interesting conundrum. Rows of blocks might wrap to the next line. Imagine you have one really big block that takes up an entire first row, and smaller blocks on the second row.

I could drag very near to the gap between smaller blocks on the second row and be slightly over the big block on the first row but nowhere near its relevant edges. I think this logic would force a drop position before or after the bigger block, but my intention would be to drop between the smaller blocks.

I think the case where this is needed is when the block is an unmodified default block. In which case `getNearestBlockIndex` might need to know whether the block is an unmodified default block.

File: packages/block-editor/src/components/use-block-drop-zone/index.js [hunk 5]
```diff
@@ -125,40 +170,58 @@ export default function useBlockDropZone( {
...
+						index: 0,
+						operation: 'insert',
+					} );
 					return;
 				}
 
-				showInsertionPoint( targetRootClientId, targetIndex );
-			}
-		}, [] ),
+				const blocks = getBlocks( targetRootClientId );
```
Comment (talldan) [inline packages/block-editor/src/components/use-block-drop-zone/index.js]: An option could be to align this with how List View drag and drop works. It assembles some metadata about blocks first - https://github.com/WordPress/gutenberg/blob/2ebd16673fa4a518dcd2149c8acd0a7632bc7d50/packages/block-editor/src/components/list-view/use-list-view-drop-zone.js#L223-L246

And then its equivalent of `getNearestBlockIndex` (`getListViewDropTarget`) uses that metadata to determine the drop target. That would make it possible to combine `getNearestBlockIndex` and `getDropTargetIndexAndOperation`. There probably wouldn't be much performance loss since `blockElements` is already filtered.

File: packages/block-editor/src/components/block-tools/insertion-point.js [hunk 6]
```diff
@@ -17,10 +17,11 @@ import { useReducedMotion } from '@wordpress/compose';
 import Inserter from '../inserter';
 import { store as blockEditorStore } from '../../store';
 import BlockPopoverInbetween from '../block-popover/inbetween';
+import BlockPopoverDropZone from '../block-popover/drop-zone';
 
 export const InsertionPointOpenRef = createContext();
 
-function InsertionPointPopover( {
+function InBetweenInsertionPointPopover( {
```
Comment (talldan) [inline packages/block-editor/src/components/block-tools/insertion-point.js]: Elsewhere in-between is considered one word (e.g. `BlockPopoverInbetween`), so it'd be good to make this consistent (`InbetweenInsertionPointPopover`).

File: packages/block-editor/src/components/block-tools/insertion-point.js [hunk 7]
```diff
@@ -229,9 +230,25 @@ function InsertionPointPopover( {
...
 
-	return isVisible && <InsertionPointPopover { ...props } />;
+	return (
+		isVisible &&
+		( insertionPoint.operation === 'replace' ? (
+			<BlockPopoverDropZone
+				// Force remount to trigger the animation.
+				key={ `${ insertionPoint.rootClientId }-${ insertionPoint.index }` }
+				{ ...props }
+			/>
```
Comment (talldan) [inline packages/block-editor/src/components/block-tools/insertion-point.js]: Nitpick: I think the component could be renamed to have the same kind of word ordering as `InBetweenInsertionPointPopover`. Maybe `BlockDropZonePopover`.

File: packages/block-editor/src/store/actions.js [hunk 8]
```diff
@@ -636,12 +636,13 @@ export function showInsertionPoint(
 	index,
 	__unstableOptions = {}
 ) {
-	const { __unstableWithInserter } = __unstableOptions;
+	const { __unstableWithInserter, operation } = __unstableOptions;
```
Comment (talldan) [inline packages/block-editor/src/store/actions.js:644]: The docs for this action could do with an update.

File: packages/block-editor/src/components/block-popover/drop-zone.js [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+import { store as blockEditorStore } from '../../store';
+import BlockPopover from './index';
+
+const animateVariants = {
+	hide: { opacity: 0, scaleY: 0.75 },
+	show: { opacity: 1, scaleY: 1 },
+	exit: { opacity: 0, scaleY: 0.9 },
+};
+
+function BlockPopoverDropZone( {
```
Comment (talldan) [inline packages/block-editor/src/components/block-popover/drop-zone.js]: It might be interesting to see whether this component can be used to solve https://github.com/WordPress/gutenberg/issues/44064.

I'd imagine it could cover the `rootClientId` block when a block list is empty.

(Should be a separate PR though)

File: packages/block-editor/src/components/use-block-drop-zone/index.js [hunk 10]
```diff
@@ -125,40 +170,58 @@ export default function useBlockDropZone( {
...
+				// eslint-disable-next-line @wordpress/no-unused-vars-before-return
+				const [ nearestBlockIndex, insertPosition ] =
+					getNearestBlockIndex(
+						blockElements,
+						{ x: event.clientX, y: event.clientY },
+						getBlockListSettings( targetRootClientId )?.orientation
+					);
+
+				// The block list is empty, don't show the insertion point but still allow dropping.
+				if ( nearestBlockIndex === undefined ) {
```
Comment (talldan) [inline packages/block-editor/src/components/use-block-drop-zone/index.js]: I wonder if this should instead be determined from the `blockElements` array or the result of `getBlocks` rather than `nearestBlockIndex`.

File: packages/block-editor/src/components/use-block-drop-zone/index.js [hunk 11]
```diff
@@ -125,40 +170,58 @@ export default function useBlockDropZone( {
...
+						blocks
+					);
+
+				setDropTarget( {
+					index: targetIndex,
+					operation,
+				} );
+				showInsertionPoint( targetRootClientId, targetIndex, {
+					operation,
+				} );
```
Comment (talldan) [inline packages/block-editor/src/components/use-block-drop-zone/index.js]: If `getDropTargetIndexAndOperation` returned an object instead of a tuple style array, it'd be possible to pass that directly to `setDropTarget`.

File: packages/block-editor/src/components/use-block-drop-zone/index.js [hunk 12]
```diff
@@ -125,40 +170,58 @@ export default function useBlockDropZone( {
...
 	);
 
-	const { getBlockListSettings } = useSelect( blockEditorStore );
+	const { getBlockListSettings, getBlocks } = useSelect( blockEditorStore );
 	const { showInsertionPoint, hideInsertionPoint } =
 		useDispatch( blockEditorStore );
 
-	const onBlockDrop = useOnBlockDrop( targetRootClientId, targetBlockIndex );
+	const onBlockDrop = useOnBlockDrop( targetRootClientId, dropTarget.index, {
+		operation: dropTarget.operation,
```
Comment (talldan) [inline packages/block-editor/src/components/use-block-drop-zone/index.js:175]: Not necessarily something to do, but thinking aloud - I guess the other option (instead of using an index for the drop target), is to use a `clientId` and then `operation` is could be 'insert-before', 'insert-after', 'insert-within' or 'replace'. I always found it hard to make a call on this as there were limited use cases at the time.

`index` was always weird because it was really the index of the gap between a block rather than the block index itself.

File: packages/block-editor/src/components/use-block-drop-zone/test/index.js (hunk 2 above)
Comment (talldan) [reply packages/block-editor/src/components/use-block-drop-zone/test/index.js:136]: Reasoning about the third block (the one with the index of 2) is useful because it's within a block list that wraps into multiple columns or rows and it's the last block in its column/row. I think it'd be ok to consider reducing the test data to only three blocks instead of removing a test case.

One thing that's nice to test is a cursor position that's outside of the block rects. `'returns "3" when the position is past the end of the third block'` previously did that. It might be worth making the test data more realistic. Like I notice all the blocks are completely adjacent, which doesn't really happen much IRL. Instead users often drag into the gap between the blocks.

**Review by talldan — COMMENTED** [review]
Looks good so far, seems like a good evolution of the feature 👍

File: packages/block-library/CHANGELOG.md (hunk 3 above)
Comment (talldan) [reply packages/block-library/CHANGELOG.md:14]: If it's possible to move it to the right place that'd be good. It's probably not super important though, I don't know how many people look at the changelog for a package like block-library.

File: packages/block-editor/src/components/use-block-drop-zone/index.js (hunk 4 above)
Comment (kevin940726) [reply packages/block-editor/src/components/use-block-drop-zone/index.js]: I don't think I fully understand what you mean here 🤔, I recorded a video and followed what you mentioned but didn't notice anything weird.

https://user-images.githubusercontent.com/7753001/193998823-105af8db-23ee-4142-b1d5-e091397f6568.mp4

Could you share more details on this?

File: packages/block-editor/src/components/use-block-drop-zone/index.js (hunk 11 above)
Comment (kevin940726) [reply packages/block-editor/src/components/use-block-drop-zone/index.js]: We would still need to extract the content to pass to `showInsertionPoint` though. And I personally think that a tuple style is more explicit and readable than returning an arbitrary object. It will not be an issue if it's fully typed though so I don't really have a strong preference 😅 .

File: packages/block-editor/src/components/use-block-drop-zone/index.js (hunk 12 above)
Comment (kevin940726) [reply packages/block-editor/src/components/use-block-drop-zone/index.js:175]: I had this idea too and I totally agree with you. That would probably be worth a bigger refactor of `useOnBlockDrop` though.

File: packages/block-editor/src/components/use-block-drop-zone/index.js (hunk 4 above)
Comment (talldan) [reply packages/block-editor/src/components/use-block-drop-zone/index.js]: Here's a couple of videos. In this PR, as soon as my cursor is slightly over the image, the drop indicator is showing before the image, even though that's not the closest place I can drop to, between '2' and '3' is still closer:

https://user-images.githubusercontent.com/677833/194002843-d1fe5929-ef66-4b0f-b126-8edf6f144adf.mp4

In trunk, this is what happens:

https://user-images.githubusercontent.com/677833/194003300-ced2691c-5a3b-432c-9eb4-9110375ff250.mp4

The flickering of the drop indicator isn't ideal, but it always chooses the nearest drop location. I think that's important because:
- We don't know what layout the user might have, the boundaries between blocks might not be very clear.
- In the video, I made the paragraphs quite big, but in reality they're usually much smaller and harder to drop between.

File: packages/block-editor/src/components/use-block-drop-zone/test/index.js (hunk 2 above)
Comment (kevin940726) [reply packages/block-editor/src/components/use-block-drop-zone/test/index.js:136]: I see! I added back the deleted tests in 86b13abd8e.

> One thing that's nice to test is a cursor position that's outside of the block rects.

Agreed! That's why I added some of these tests in the e2e tests below. It would be good to review them as well.

File: packages/block-library/CHANGELOG.md (hunk 3 above)
Comment (kevin940726) [reply packages/block-library/CHANGELOG.md:14]: Yeah, I moved the changelog to `block-editor`.

File: packages/block-editor/src/components/use-block-drop-zone/index.js (hunk 4 above)
Comment (kevin940726) [reply packages/block-editor/src/components/use-block-drop-zone/index.js]: I see! I'll see what I can do.

Comment (kevin940726) [reply packages/block-editor/src/components/use-block-drop-zone/index.js]: I discovered one issue when trying to fix this. Why are we using `blockElements` and getting them manually rather than using `getBlocks()`? I found out that `blockElements` will contain the default block appender as well. I'm not sure when we'll want that?

Comment (talldan) [reply packages/block-editor/src/components/use-block-drop-zone/index.js]: `getBlocks()` is only block data, and the code needs the DOM element's bounding rect. It's certainly not ideal how it relies on the DOM particulars of blocks and the block list.

I know there's now `useBlockElement` (it didn't exist at the time of writing this code), so it may now be possible to get the elements for an entire block list using more of an editor API. (maybe adding a new `useBlockListElements( rootClientId )` is possible?)

If we have to do `getElementById` or `querySelector` in a map, then it'd definitely be worth checking on performance. That feels potentially expensive.

Comment (kevin940726) [reply packages/block-editor/src/components/use-block-drop-zone/index.js]: I think it should be pretty fast to do `document.getElementById`. Calling it 1000000 times in a loop only spends around 30ms on my machine.

File: packages/block-editor/src/components/use-block-drop-zone/index.js (hunk 5 above)
Comment (kevin940726) [reply packages/block-editor/src/components/use-block-drop-zone/index.js]: Did a version of that in [cce8d8f](https://github.com/WordPress/gutenberg/pull/44647/commits/cce8d8f56a97d334f1f5e8381f2f7ee80bfee311).

File: packages/block-editor/src/components/use-block-drop-zone/index.js (hunk 4 above)
Comment (kevin940726) [reply packages/block-editor/src/components/use-block-drop-zone/index.js]: I ended up refactored it to `getDropTargetPosition` in cce8d8f.

File: packages/block-editor/src/components/use-block-drop-zone/test/index.js (hunk 2 above)
Comment (kevin940726) [reply packages/block-editor/src/components/use-block-drop-zone/test/index.js:136]: I refactored the tests again in cce8d8f.

**Comment by kevin940726** [comment]
Pinging @ellatrix and @youknowriad in case you missed it ;)

File: packages/block-editor/src/components/block-tools/insertion-point.js [hunk 13]
```diff
@@ -229,9 +230,25 @@ function InsertionPointPopover( {
...
+	return (
+		isVisible &&
+		( insertionPoint.operation === 'replace' ? (
+			<BlockDropZonePopover
+				// Force remount to trigger the animation.
+				key={ `${ insertionPoint.rootClientId }-${ insertionPoint.index }` }
+				{ ...props }
+			/>
+		) : (
+			<InbetweenInsertionPointPopover { ...props } />
```
Comment (youknowriad) [inline packages/block-editor/src/components/block-tools/insertion-point.js]: I think right now this is ignoring the "isVisible" check right? Should we return early if "! isVisible" instead?

**Comment by youknowriad** [comment]
https://user-images.githubusercontent.com/272444/194242179-ad506cfb-23ca-4de8-b58f-4e124a7693ca.mov

I'm noticing two issues in the following recording:

 - Dragging over the paragraph blocks flickers the "replace" insertion point too much. I suspect that we're calling showInsertionPoint then hideInsertionPoint too much.
 - I expected that if I hover the area between two paragraphs, that I'd be able to insert at that position without replacing the preceding empty paragraph

File: packages/block-editor/src/components/use-block-drop-zone/index.js [hunk 14]
```diff
@@ -25,66 +30,99 @@ import { store as blockEditorStore } from '../../store';
...
+				rect,
+				allowedEdges
+			);
+			// Prioritize the element if the point is inside of an unmodified default block.
+			if (
+				isUnmodifiedDefaultBlock &&
+				isPointContainedByRect( position, rect )
+			) {
+				distance = 0;
+			}
```
Comment (talldan) [inline packages/block-editor/src/components/use-block-drop-zone/index.js:84]: There's probably a chance to break the loop early here.

File: packages/block-editor/src/components/block-tools/insertion-point.js (hunk 13 above)
Comment (kevin940726) [reply packages/block-editor/src/components/block-tools/insertion-point.js]: I think `isVisible` is at [L243](https://github.com/WordPress/gutenberg/pull/44647/files/43b266eaa21f322a688e1dc39fccb08c80aad5ee#diff-5a88c3390c631539a539a69de522717ef1e0fc69900524a458dd225b20c6b2aeR243)?

**Comment by kevin940726** [comment]
Thanks for trying it out!

> Dragging over the paragraph blocks flickers the "replace" insertion point too much. I suspect that we're calling showInsertionPoint then hideInsertionPoint too much.

Are you testing it in Safari? There's a [known bug](https://github.com/WordPress/gutenberg/pull/44647#discussion_r986426064) in Safari that I think we should fix in a follow-up.

>  I expected that if I hover the area between two paragraphs, that I'd be able to insert at that position without replacing the preceding empty paragraph

That's intentional though. From the [suggestion](https://github.com/WordPress/gutenberg/pull/42722#issuecomment-1236082016) in the original PR, I think the idea is to avoid showing the in-between indicator when there's an empty paragraph block because the outcome will be the same. That's up for debate though.

However, it reminds me of the indicator while hovering in the global inserter. I think we should also change that in a follow-up PR (if that's expected). 🤔

File: packages/block-editor/src/components/block-tools/insertion-point.js (hunk 13 above)
Comment (youknowriad) [reply packages/block-editor/src/components/block-tools/insertion-point.js]: You're right I mis-read the parenthesis.

**Comment by youknowriad** [comment]
> Are you testing it in Safari? There's a https://github.com/WordPress/gutenberg/pull/44647#discussion_r986426064 in Safari

Yes,

> That's intentional though. From the https://github.com/WordPress/gutenberg/pull/42722#issuecomment-1236082016 in the original PR, I think the idea is to avoid showing the in-between indicator when there's an empty paragraph block because the outcome will be the same. That's up for debate though.

I guess my point is that the outcome shouldn't be the same, I should leave the empty paragraph and insert after it. At least in my testing it felt weird that the "blue area" was visible when I was clearly not on top of it. cc @mtias

**Comment by kevin940726** [comment]
> At least in my testing it felt weird that the "blue area" was visible when I was clearly not on top of it.

I think it makes sense if you think of the dropzone as a variant of the blue line indicator. It indicates where the item should be placed. It feels a bit weird to me if you can't drop in between empty paragraphs while you can do that for other block types. Maybe we can come up with a better design though, maybe even in a different issue?

**Comment by youknowriad** [comment]
> It feels a bit weird to me if you can't drop in between empty paragraphs while you can do that for other block types. 

Yeah, that's my feeling too. Anyway, if this is how it already works in trunk, we can track it separately.

File: packages/block-editor/src/components/block-tools/insertion-point.js [hunk 15]
```diff
@@ -229,9 +230,25 @@ function InsertionPointPopover( {
...
+		return {
+			insertionPoint: getBlockInsertionPoint(),
+			isVisible: isBlockInsertionPointVisible(),
+		};
 	}, [] );
 
-	return isVisible && <InsertionPointPopover { ...props } />;
+	return (
+		isVisible &&
+		( insertionPoint.operation === 'replace' ? (
```
Comment (talldan) [inline packages/block-editor/src/components/block-tools/insertion-point.js]: It could be worth adding a small comment here along the lines of "When replacing, use a popover that overlays the block, and when inserting use a popover that appears between blocks."

File: packages/block-editor/src/components/use-block-drop-zone/index.js (hunk 12 above)
Comment (talldan) [reply packages/block-editor/src/components/use-block-drop-zone/index.js:175]: I guess backwards compatibility will be difficult 🤔

File: packages/block-editor/src/components/use-block-drop-zone/index.js [hunk 16]
```diff
@@ -125,40 +166,58 @@ export default function useBlockDropZone( {
...
+					getBlockListSettings( targetRootClientId )?.orientation
+				);
+
+				setDropTarget( {
+					index: targetIndex,
+					operation,
+				} );
+				showInsertionPoint( targetRootClientId, targetIndex, {
+					operation,
+				} );
```
Comment (talldan) [inline packages/block-editor/src/components/use-block-drop-zone/index.js:217]: This is subjective, but is `operation: 'insert' | 'replace'` is the right terminology for the action of `showInsertionPoint`?

This action displays the UI, so I wonder if the option should be something like `appearance: 'inbetween' | 'overlay-block'`. That might also better match the terminology in the `InsertionPoint` component.

`useOnBlockDrop` could still keep the insert/replace terms because it does seem like the right thing there.

I guess the only thing then is becomes a bit of an overhead to translate `insert` to `inbetween` and `replace` to `overlay-block`.

The other thing is that it is called 'insertion' point, which implies 'insert' already. 🤔 

I am nitpicking now, so feel free to ignore 😄

File: packages/block-editor/src/components/use-block-drop-zone/index.js [hunk 17]
```diff
@@ -125,40 +166,58 @@ export default function useBlockDropZone( {
...
+				const blocksData = blocks.map( ( block ) => {
+					const clientId = block.clientId;
+
+					return {
+						isUnmodifiedDefaultBlock:
+							getIsUnmodifiedDefaultBlock( block ),
+						getBoundingClientRect: () =>
+							ownerDocument
+								.getElementById( `block-${ clientId }` )
+								.getBoundingClientRect(),
```
Comment (talldan) [inline packages/block-editor/src/components/use-block-drop-zone/index.js:200]: That is nicer, and it also ensures the blocks don't need to be children of the current element, so should make things more stable. Good improvement 👍

**Review by talldan — APPROVED** [review]
LGTM, I left a few comments, but nothing blocking.

Great work on refactoring this feature!

File: packages/block-editor/src/components/use-block-drop-zone/index.js (hunk 14 above)
Comment (kevin940726) [reply packages/block-editor/src/components/use-block-drop-zone/index.js:84]: I had this plan to use binary search to find the index. Better to do them together in another PR I guess ;)

**Comment by ellatrix** [comment]
Please allow me to have a look at this PR today.

**Comment by ellatrix** [comment]
> This refactor allows us to reuse the same drop zone and their event listeners so that we don't have to bind them elsewhere.

Could you elaborate on this? Why is this needed?

File: packages/e2e-test-utils-playwright/src/page-utils/drag-files.ts [hunk 18]
```diff
@@ -99,14 +100,19 @@ async function dragFiles(
...
-		 * @param  options.position A point to use relative to the top-left corner of element padding box. If not specified, uses some visible point of the element.
+		 * @param  selectorOrLocator A selector or a locator to search for an element.
+		 * @param  options           The optional options.
+		 * @param  options.position  A point to use relative to the top-left corner of element padding box. If not specified, uses some visible point of the element.
 		 */
-		dragOver: async ( selector: string, options: Options = {} ) => {
-			const boundingBox = await this.page
-				.locator( selector )
-				.boundingBox();
+		dragOver: async (
```
Comment (ellatrix) [inline packages/e2e-test-utils-playwright/src/page-utils/drag-files.ts:107]: Why are we using Playwright here instead of Puppeteer if in Puppeteer drag and drop works out of the box?

**Comment by ellatrix** [comment]
I like this way more than the previous implementation! 🙂

When I test dragging a block, I get the following React warning:

Screenshot 2022-10-12 at 15 02 49

File: packages/e2e-test-utils-playwright/src/page-utils/drag-files.ts (hunk 18 above)
Comment (kevin940726) [reply packages/e2e-test-utils-playwright/src/page-utils/drag-files.ts:107]: Several reasons:

1. We're in the middle of the migration from Puppeteer to Playwright. In order to make the migration path smoother, it makes sense for me to add some missing features in `e2e-test-utils-playwright`. Once these utils are implemented, I believe it will be easier for contributors to pick up or add more tests.
2. Puppeteer doesn't fully support drag-and-drop out of the box either. Specifically, if I'm not mistaken, Puppeteer also doesn't support API for dragging files (https://github.com/puppeteer/puppeteer/issues/7330). It seems like changes need to be made in Chrome Devtools Protocol to support this. If we implement it in Puppeteer, we would still have to implement these utils anyway.

@talldan has opened an issue (https://github.com/microsoft/playwright/issues/16437) in the Playwright repo already, I believe that's a good start. We can iterate and see where it goes from there ;)

**Comment by kevin940726** [comment]
> Could you elaborate on this? Why is this needed?

This is a cleaner implementation as we don't have to lock in to the Paragraph block or any other blocks. The dropzone behaves very similarly to the insertion point, I think it makes sense to combine them both and share common logic. This also allows us to avoid binding event listeners for each dropzone, this could be a performance bottleneck if we render tons of dropzones at the same time.

> When I test dragging a block, I get the following React warning:

I believe this is an existing bug in the `<Tooltip>` component as I didn't touch those in this PR. I can still reproduce it in a commit even before #42722.

File: packages/block-editor/src/components/use-block-drop-zone/index.js (hunk 16 above)
Comment (kevin940726) [reply packages/block-editor/src/components/use-block-drop-zone/index.js:217]: Yeah, makes sense. It's a bit unfortunate that we're probably too late to change the naming of `insertion point` now 😅 .

**Review by ellatrix — APPROVED** [review]

File: packages/block-editor/src/components/block-tools/insertion-point.js (hunk 15 above)
Comment (kevin940726) [reply packages/block-editor/src/components/block-tools/insertion-point.js]: Added in [b4daabc](https://github.com/WordPress/gutenberg/pull/44647/commits/b4daabc10826525d92b6b32d889cf760a41fcddc)!

**Comment by youknowriad** [comment]
Thanks for iterating on this :)
